rust/actyx/ax-aql/src/language/parser.rs

```diff
@@ -721,6 +722,116 @@ pub(crate) fn query_from_str(s: &str) -> Result<Query<'_>> {
     Ok(query)
 }
 
+pub(crate) struct QueryWorkflowAnalysisError<'a> {
+    pub(crate) span: Span<'a, Ident>,
+    pub(crate) error_type: QueryWorkflowAnalysisErrorType,
+}
+
+pub(crate) enum QueryWorkflowAnalysisErrorType {
+    UndeclaredEvent,
+    UndeclaredWorkflowCall,
+}
+
+pub(crate) struct QueryWorkflowAnalysis<'a> {
+    pub(crate) query: &'a Query<'a>,
+    pub(crate) workflow_tracker: WorkflowTracker<'a>,
+}
+
+impl<'a> QueryWorkflowAnalysis<'a> {
+    pub fn from(query: &'a Query) -> Self {
+        Self {
+            query,
+            workflow_tracker: Default::default(),
+        }
+    }
+
+    pub fn check(&'a self, query: &'a Query) -> Vec<QueryWorkflowAnalysisError> {
+        let mut errors = Vec::new();
+        query
+            .workflows
+            .iter()
+            .for_each(|(_ident, workflow)| errors.extend(self.check_workflow(workflow)));
+        errors
+    }
+
+    fn check_workflow(&'a self, workflow: &'a Workflow) -> Vec<QueryWorkflowAnalysisError> {
+        if self.workflow_tracker.contains(workflow) {
+            return Vec::new();
+        }
+        self.workflow_tracker.insert(workflow);
+        self.check_steps(&workflow.steps)
+    }
+
+    fn check_steps(&'a self, steps: &'a NonEmptyVec<WorkflowStep<'a>>) -> Vec<QueryWorkflowAnalysisError> {
+        let mut errors = Vec::new();
+        steps.iter().for_each(|step| errors.extend(self.check_step(step)));
+        errors
+    }
+
+    fn check_step(&'a self, step: &'a WorkflowStep<'a>) -> Vec<QueryWorkflowAnalysisError> {
+        use super::workflow::WorkflowStep::*;
+
+        match step {
+            Event { label, binders: _, .. } => {
+                let mut errors = Vec::new();
+                if !self.query.events.contains_key(&label) {
+                    errors.push(QueryWorkflowAnalysisError {
+                        error_type: QueryWorkflowAnalysisErrorType::UndeclaredEvent,
+                        span: label.clone(),
+                    });
+                }
+                // this part will also handle additional stuffs like type matching in `binders`
+                errors
+            }
+            Retry { steps } => self.check_steps(steps),
+            Timeout { steps, .. } => self.check_steps(steps),
+            Parallel { cases, .. } => {
+                let mut errors = Vec::new();
+                cases.iter().for_each(|case| errors.extend(self.check_steps(case)));
+                errors
+            }
+            Call {
+                workflow: workflow_ident,
+                cases: _,
+                ..
+            } => {
+                if let Some(workflow) = self.query.workflows.get(&*workflow_ident) {
+                    self.check_workflow(workflow)
```

Comment:
I wouldn’t do this here, since we have a reliable top-level enumeration of all workflows. This means the tracker isn’t needed.